### PR TITLE
test cases  for progress_bar.py

### PR DESCRIPTION
**Pull request by VickyShel**

### CODE REVIEW

File: .idea/.gitignore
Comment (fostiropoulos) [inline .idea/.gitignore:1]: Is this file necessary?

File: tests/utils/test_progress_bar.py [hunk 1]
```diff
@@ -134,6 +134,63 @@ def test_remote(i, progress_bar: RemoteProgressBar):
...
+        result=in_notebook()
+        assert result==False
+
+from ablator.utils.progress_bar import get_last_line
+def test_get_last_line():
+    result=get_last_line(Path("hhhh.txt"))
+    assert result==None
+    result=get_last_line(None)
+    assert result==None
+    result=get_last_line(Path("/Users/vivi/Documents/USC/实习/Ablator/ablator_v0.0.1-mp/test2.txt"))
```
Comment (fostiropoulos) [inline tests/utils/test_progress_bar.py]: Is this path available to all users?

**Review by fostiropoulos — CHANGES_REQUESTED** [review]
Requires several revisions.

**Comment by fostiropoulos** [comment]
@apneetha can you please review?

**Comment by apneetha** [comment]
Sure, I'll review it 

> @apneetha can you please review?

**Comment by apneetha** [comment]
@fostiropoulos 

Hey  @VickyShel   ,  
As the one of the path for the test_get_line() is not available to all users, you could write a test case by creating a temporary file 
Instead of a passing None to the get_last_line(), you could pass an empty file.


```
def test_get_last_line(tmpdir):
    #Create a temporary file
    test_file = Path(tmpdir.join("test_file.txt"))

    # Test cases
    lines = [
        "This is the first line.",
        "Here's another line.",
        "This is the last line."
    ]
    
    with open(test_file, 'w') as f:
        f.write('\n'.join(lines))

    # Check if the function returns the expected last line
    assert get_last_line(test_file) == "This is the last line."

    # Edge case: empty file
    empty_test_file = Path(tmpdir.join("empty_test_file.txt"))
    empty_test_file.touch()  # creates an empty file
   
    try:
       result = get_last_line(empty_test_file)
    except Exception as e:
        assert False, f"Function raised an exception for an empty file: {e}"
```


```
def test_get_last_line_single_line_without_newline(tmpdir):
    # Create a temporary file
    test_file = Path(tmpdir.join("single_line_file.txt"))

    # Test case with a single line not ending with a newline
    line = "This is the only line in the file, and it doesn't end with a newline."
    
    with open(test_file, 'w') as f:
        f.write(line)

    # Check if the function returns the expected line
    assert get_last_line(test_file) == line
```

**Comment by apneetha** [comment]
@VickyShel ,

For the def test_display_class(),
Could you write separate test case for each of the following, instead of having them as one pytest. 

#test _display function of Display class
#test _refresh function of Display class
#test _update_screen_dims function of Display class
#test print_texts function of Display class

File: tests/utils/test_progress_bar.py [hunk 2]
```diff
@@ -134,6 +134,63 @@ def test_remote(i, progress_bar: RemoteProgressBar):
         dis.refresh()
         time.sleep(random.random() / 10)
 
+from ablator.utils.progress_bar import in_notebook
+def test_in_notebook():
+    result=in_notebook()
+    assert result==False,"The in_notebook function cannot correctly determine whether it is a terminal."
+    with mock.patch('ablator.utils.progress_bar.in_notebook', side_effect=ImportError):
```
Comment (fostiropoulos) [inline tests/utils/test_progress_bar.py]: I do not think this is a good test to see if it is robust in Notebook. The import errors could be related to other issues 
```
if "IPKernelApp" not in get_ipython().config:  # pragma: no cover
    return False
```

we should find a way to simulate that the ipykernel is running. Will require significant research.

File: tests/utils/test_progress_bar.py [hunk 3]
```diff
@@ -134,6 +134,63 @@ def test_remote(i, progress_bar: RemoteProgressBar):
...
+        result = in_notebook()
+        assert result == False
+    with mock.patch('ablator.utils.progress_bar.in_notebook',side_effect=AttributeError):
+        result=in_notebook()
+        assert result==False
+
+from ablator.utils.progress_bar import get_last_line
```
Comment (fostiropoulos) [inline tests/utils/test_progress_bar.py]: imports must be on top-level

**Review by fostiropoulos — CHANGES_REQUESTED** [review]

File: .idea/.gitignore
Comment (VickyShel) [reply .idea/.gitignore:1]: I will commit a new PR later. Sorry for the inconvenience.

**Comment by fostiropoulos** [comment]
changes have been merged

**Pull request closed without merging**